pkg/client/client_v2.go

```diff
@@ -365,42 +393,90 @@ func (c *clientV2) syncUnits(expected *proto.CheckinExpected) {
 			c.units[i] = unit
 			i++
 		} else {
-			c.unitsCh <- UnitChanged{
+			changed := UnitChanged{
 				Type: UnitChangedRemoved,
 				Unit: unit,
 			}
+
+			changed = c.syncFeatures(expected, changed)
+
+			c.changesCh <- changed
 			removed = true
 		}
 	}
+
 	// resize so units that no longer exist are removed from the slice
 	c.units = c.units[:i]
 	for _, agentUnit := range expected.Units {
 		unit := c.findUnit(agentUnit.Id, UnitType(agentUnit.Type))
 		if unit == nil {
 			// new unit
-			unit = newUnit(agentUnit.Id, UnitType(agentUnit.Type), UnitState(agentUnit.State), UnitLogLevel(agentUnit.LogLevel), agentUnit.Config, agentUnit.ConfigStateIdx, c)
+			unit = newUnit(
+				agentUnit.Id,
+				UnitType(agentUnit.Type),
+				UnitState(agentUnit.State),
+				UnitLogLevel(agentUnit.LogLevel),
+				agentUnit.Config,
+				agentUnit.ConfigStateIdx,
+				c)
 			c.units = append(c.units, unit)
-			c.unitsCh <- UnitChanged{
+
+			changed := UnitChanged{
 				Type: UnitChangedAdded,
 				Unit: unit,
 			}
+
+			changed = c.syncFeatures(expected, changed)
+
+			c.changesCh <- changed
 		} else {
 			// existing unit
-			if unit.updateState(UnitState(agentUnit.State), UnitLogLevel(agentUnit.LogLevel), agentUnit.Config, agentUnit.ConfigStateIdx) {
-				c.unitsCh <- UnitChanged{
-					Type: UnitChangedModified,
-					Unit: unit,
+			triggers := unit.updateState(
+				UnitState(agentUnit.State),
+				UnitLogLevel(agentUnit.LogLevel),
+				agentUnit.Config,
+				agentUnit.ConfigStateIdx)
+			if triggers != nil { // a.k.a something changed
+				changed := UnitChanged{
+					Triggers: triggers,
+					Type:     UnitChangedModified,
+					Unit:     unit,
 				}
+
+				changed = c.syncFeatures(expected, changed)
```

Comment:
This should be updating the triggers list shouldn't it?

Also since Features exist at the top level of the protocol and is not actually per unit, it feels like this should happen outside the `for _, agentUnit := range expected.Units {` loop above and then the result and trigger (if needed) copied into each unit.

Otherwise I suspect you'd only detect a feature change for the first unit in this list, and omit the trigger from the others unless you kept a copy of the Features for each unit individually.